### PR TITLE
Report missing cases of `bare_trait_objects`

### LINKED ISSUE #65371: Lint `bare_trait_objects` doesn't warn on traits used as types in paths
Example:
```rust
trait Method { fn method() {} }

trait Tr {}

impl Method for Tr {} // warning: trait objects without an explicit `dyn` are deprecated

fn main() {
    Tr::method; // No warning
}
```
According to the logic of the lint there should be a suggestion to rewrite `Tr::method` as `<dyn Tr>::method`.

### PR DESCRIPTION
**Pull request by petrochenkov**
Fixes https://github.com/rust-lang/rust/issues/65371

### CODE REVIEW

File: src/test/ui/unspecified-self-in-trait-ref.stderr [hunk 1]
```diff
@@ -1,29 +1,61 @@
+warning: trait objects without an explicit `dyn` are deprecated
+  --> $DIR/unspecified-self-in-trait-ref.rs:10:13
+   |
+LL |     let a = Foo::lol();
+   |             ^^^^^^^^ help: use `dyn`: `dyn Foo::lol`
```
Comment (estebank) [inline src/test/ui/unspecified-self-in-trait-ref.stderr:5]: The suggestions are now wrong, right? These should be `<dyn Foo>::lol`, right?

**Comment by estebank** [comment]
> It should probably be reported later instead, only if the path ends up being successfully resolved.

`maybe_lint_bare_trait` is already using `buffered_lints`, but they are still emitted after parse. We could potentially delay them further and prune the list during resolve as appropriate.

**Comment by estebank** [comment]
Beyond the concerns about now incorrect suggestions and the increased number of warnings for code in the wild, r=me. If you land this without addressing the two comments, please file a follow up ticket so we don't forget to fix them eventually.

**Comment by petrochenkov** [comment]
>We could potentially delay them further and prune the list during resolve as appropriate.

It has to be delayed until type checking when type-relative associated items are resolved.
I'll try to look at this tomorrow.

**Comment by petrochenkov** [comment]
Blocked on https://github.com/rust-lang/rust/pull/83092.

**Comment by petrochenkov** [comment]
Updated.

**Comment by estebank** [comment]
r=me after fixing clippy

**Pull request merged**